src/ray/common/ray_syncer/ray_syncer-inl.h

```diff
@@ -0,0 +1,201 @@
+  const absl::flat_hash_map<std::string, Array<std::shared_ptr<const RaySyncMessage>>>
+      &GetClusterView() const {
+    return cluster_view_;
+  }
+
+ private:
+  /// For local nodes
+  Array<const ReporterInterface *> reporters_ = {nullptr};
+  Array<ReceiverInterface *> receivers_ = {nullptr};
+
+  /// This field records the version of the snapshot that has been taken.
+  Array<int64_t> snapshots_taken_;
+  /// Keep track of the latest messages received.
+  /// Use shared pointer for easier liveness management since these messages might be
+  /// sending via rpc.
+  absl::flat_hash_map<std::string, Array<std::shared_ptr<const RaySyncMessage>>>
+      cluster_view_;
+};
+
+class NodeSyncConnection {
+ public:
+  NodeSyncConnection(RaySyncer &instance,
+                     instrumented_io_context &io_context,
+                     std::string node_id);
+
+  /// Push a message to the sending queue to be sent later. Some message
+  /// might be dropped if the module think the target node has already got the
+  /// information. Usually it'll happen when the message has the same node id
+  /// as the target or the message is sent from this node.
+  ///
+  /// \param message The message to be sent.
+  ///
+  /// \return true if push to queue successfully.
+  bool PushToSendingQueue(std::shared_ptr<const RaySyncMessage> message);
+
+  /// Send the message queued.
+  virtual void DoSend() = 0;
+
+  virtual ~NodeSyncConnection() {}
+
+  /// Return the node id of this sync context.
+  const std::string &GetNodeId() const { return node_id_; }
+
+  /// Handle the udpates sent from this node.
+  ///
+  /// \param messages The message received.
+  void ReceiveUpdate(RaySyncMessages messages);
+
+ protected:
+  FRIEND_TEST(RaySyncerTest, NodeSyncConnection);
+
+  std::array<int64_t, kComponentArraySize> &GetNodeComponentVersions(
+      const std::string &node_id);
+
+  RaySyncer &instance_;
+  instrumented_io_context &io_context_;
+  std::string node_id_;
+
+  struct _MessageHash {
+    std::size_t operator()(
+        const std::shared_ptr<const RaySyncMessage> &m) const noexcept {
+      std::size_t seed = 0;
+      boost::hash_combine(seed, m->node_id());
+      boost::hash_combine(seed, m->component_id());
+      return seed;
+    }
+  };
+
+  struct _MessageEq {
+    bool operator()(const std::shared_ptr<const RaySyncMessage> &lhs,
+                    const std::shared_ptr<const RaySyncMessage> &rhs) const noexcept {
+      if (lhs == rhs) {
+        return true;
+      }
+      if (lhs == nullptr || rhs == nullptr) {
+        return false;
+      }
+      // We don't check the version here since we want the old version to be deleted.
+      return lhs->node_id() == rhs->node_id() &&
+             lhs->component_id() == rhs->component_id();
+    }
+  };
+
+  absl::flat_hash_set<std::shared_ptr<const RaySyncMessage>, _MessageHash, _MessageEq>
+      sending_queue_;
+  // Keep track of the versions of components in this node.
```

Comment:
it's not (just) about components in this node right? Also components in other nodes?

Reply:
I'm still a bit confused: is this about the latest version the local node has received or sent?

Reply:
Now I understand what this hash map is, but I think we should still expand the comment to have more details: e.g. examples of what's stored there and how they will be used?